pkg/controller/statefulset/stateful_set_control.go

```diff
@@ -404,6 +404,21 @@ func (ssc *defaultStatefulSetControl) updateStatefulSet(
 				replicas[i].Name)
 			return &status, nil
 		}
+		// Enforce the StatefulSet invariants - we do this without respect to the Pod's readiness so that the endpoints
+		// controller can be notified of identity changes if a Pod becomes unready due to a DNS inconsistency with respect
+		// to the Pods identity.
+		if !identityMatches(set, replicas[i]) || !storageMatches(set, replicas[i]) {
```

Comment:
I tried a manual test of this and found a bug. We also need to put the check for hostname/subdomain back into `identityMatches()`, or else this doesn't trigger.